docs/src/explanations/dpi.md

```diff
@@ -0,0 +1,143 @@
+| 64    | longint      | value                  |
+| > 64  | bit [w-1:0]  | reference              |
+
+### Function Types
+The type of DPI object you need depends on the characteristics of your C++ function:
+
+* Return Type
+  * For functions that don't return a value (like hello), use `DPIVoidFunctionImport`.
+  * For functions with a return type (e.g., integer addition), use `DPINonVoidFunctionImport[T]`, where `T` is the return type.
+  * The output argument must be the last argument in DPI fuction.
+* Clocked vs. Unclocked:
+  * Clocked: The function call is evaluated at the associated clock's positive edge. By default, Chisel uses the current module's clock. For custom clocks, use `withClocked(clock) {..}`. If the function has a return value, it will be available in the next clock cycle.
+  * Unclocked: The function call is evaluated immediately.
+
+## Example: Adding Two Numbers
+Here's an example of a DPI function that calculates the sum of two numbers:
+
+```c++
+extern "C" void add(int lhs, int rhs, int* result)
+{
+    *result = lhs + rhs;
+}
+```
+
+```scala
+object Add extends DPINonVoidFunctionImport[UInt] {
+  override val functionName = "add"
+  override val ret = UInt(32.W)
+  override val clocked = false
+  override val inputNames = Some(Seq("lhs", "rhs"))
+  override val outputName = Some("result")
+  final def apply(lhs: UInt, rhs: UInt): UInt = super.call(lhs, rhs)
+}
+
+val result = Add(2.U(32.W), 3.U(32.W))
+```
+
+Explanation:
+
+* `Add` inherits from `DPINonVoidFunctionImport[UInt]` because it returns a 32-bit unsigned integer.
+* `ret` specifies the return type.
+* `clocked` indicates that this is a clocked function call.
+* `inputNames` and `outputName` provide optional names for the function's arguments and return value (these are just for Verilog readability).
+
+## Example: Sum of an array
+Chisel vectors are converted into SystemVerilog open arrays when used with DPI. Since memory layout can vary between simulators, it's recommended to use `svSize` and `svGetBitArrElemVecVal` to access array elements.
+
+```c++
+extern "C" void sum(const svOpenArrayHandle array, int* result) {
+  // Get a length of the open array.
+  int size = svSize(array, 1);
+  // Initialize the result value.
+  *result = 0;
+  for(size_t i = 0; i < size; ++i) {
+    svBitVecVal vec;
+    svGetBitArrElemVecVal(&vec, array, i);
+    *result += vec;
+  }
+}
+```
+
+```scala
+object Sum extends DPINonVoidFunctionImport[UInt] {
+  override val functionName = "sum"
+  override val ret = UInt(32.W)
+  override val clocked = false
+  override val inputNames = Some(Seq("array"))
+  override val outputName = Some("result")
+  final def apply(array: Vec[UInt]): UInt = super.call(array)
+}
+
+val io = IO(new Bundle {
+  val a = Input(Vec(3, UInt(32.W)))
+  val b = Input(Vec(6, UInt(32.W)))
+})
+
+val sum_a = Sum(io.a) // compute a[0] + a[1] + a[2]
+val sum_b = Sum(io.b) // compute b[0] + ... + b[5]
+```
+
+# FAQ
+
```

Comment:
can you show an example where having clocked = false makes sense? Is that done in a `when` context...?

Reply:
Unclocked calls be useful when user wants to replace pure (but expensive) combinatorial logics with dpi call (actually `Add` is a good example of such use case). 
An unclocked call under `when(cond)` is lowered into `always_comb` + `if(cond)` and DPI is conditionally invoked as well. I updated `Add` example to include both of them. 

```scala
class AddTest extends Module {
  val io = IO(new Bundle {
    val a = Input(UInt(32.W))
    val b = Input(UInt(32.W))
    val c = Output(UInt(32.W))
    val d = Output(UInt(32.W))
    val en = Input(Bool())
  })

  // Call DPI only when `en` is true.
  when (io.en) {
    io.c := AddClocked(io.a, io.b)
    io.d := AddUnclocked(io.a, io.b)
  } .otherwise {
    io.c := 0.U(32.W)
    io.d := 0.U(32.W)
  }
}
```
```verilog
module AddTest(
  input         clock,
                reset,
  input  [31:0] io_a,
                io_b,
  output [31:0] io_c,
                io_d,
  input         io_en
);

  logic [31:0] _add_0;
  reg   [31:0] _GEN;
  always @(posedge clock) begin
    if (io_en) begin
      add(io_a, io_b, _add_0);
      _GEN <= _add_0;
    end
  end // always @(posedge)
  reg   [31:0] _GEN_0;
  always_comb begin
    if (io_en) begin
      add(io_a, io_b, _GEN_0);
    end
    else
      _GEN_0 = 32'bx;
  end // always_comb
  assign io_c = io_en ? _GEN : 32'h0;
  assign io_d = io_en ? _GEN_0 : 32'h0;
```